core/src/main/java/ai/timefold/solver/core/impl/util/LinkedIdentityHashSet.java

```diff
@@ -0,0 +1,77 @@
+package ai.timefold.solver.core.impl.util;
+
+import java.util.AbstractSet;
+import java.util.IdentityHashMap;
+import java.util.Iterator;
+
+import org.jspecify.annotations.NullMarked;
+
+/**
+ * This set does not support null keys.
+ * 
+ * @param <V>
+ */
+@NullMarked
+public final class LinkedIdentityHashSet<V> extends AbstractSet<V> {
+
+    private final ElementAwareList<V> delegate;
+    private final IdentityHashMap<V, ElementAwareListEntry<V>> identityMap;
+    private int size = 0; // Avoid method calls to underlying collections.
+
+    public LinkedIdentityHashSet() {
+        this.delegate = new ElementAwareList<>();
+        this.identityMap = new IdentityHashMap<>();
+    }
+
+    @Override
+    public Iterator<V> iterator() {
+        return delegate.iterator();
+    }
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        if (size == 0) { // Micro-optimization; contains() on an empty map is not entirely free.
+            return false;
+        }
+        return identityMap.containsKey(o);
+    }
+
+    @Override
+    public boolean add(V v) {
+        var entry = identityMap.get(v);
+        if (entry == null) {
+            identityMap.put(v, delegate.add(v));
+            size += 1;
+            return true;
+        }
```

Comment:
computeIfAbsent does the same thing, but it also forces the JVM to create a lambda instance. This does not.